# Komplizierte JSONs auslesen

In [12]:
import pandas as pd
import numpy as np
import glob
import json
%matplotlib inline

In [13]:
path = 'psyreg_output/'
filenames = glob.glob(path + '*.json')

In [14]:
len(filenames)

2

In [15]:
data = []
with open(filenames[0]) as json_data:
    data = json.load(json_data)
data

{'idPerson': 1,
 'name': 'Birchler Pedross',
 'vorname': 'Angelina',
 'geschlechtId': 12001,
 'geburtsdatum': None,
 'jahrgang': 1968,
 'sprachkompetenzIds': [13000],
 'nationalitaet1Id': 10001,
 'nationalitaet2Id': None,
 'gln': '7601003917000',
 'uid': None,
 'adressenBewilligungen': [{'kantonId': 11020,
   'praxisname': '-',
   'strasse': 'Zelgstr. 5',
   'plz': '8003',
   'ort': 'Zürich',
   'telefon1': '',
   'emailadresse': '',
   'zusatz1': '',
   'zusatz2': ''}],
 'adressenNeunzigTageDienstleister': [],
 'bewilligungenErteilt': [{'kantonId': 11020,
   'auflageEinschraenkungen': [{'typId': 8003,
     'datumEntscheid': '2010-08-02T00:00:00',
     'datumBefristung': '2020-08-31T00:00:00'}],
   'datumEntscheid': '2010-08-02T00:00:00',
   'datumErteilung': '2018-10-23T15:12:51.923',
   'datumBefristung': '2018-10-30T07:07:15.607'}],
 'bewilligungenEntzogen': [],
 'bewilligungenVerweigert': [],
 'bewilligungenKeineBewilligung': [],
 'neunzigTageDienstleister': [],
 'abschluesse': [{'

Eine der verschachtelten Listen von Dicts ist zum Beispiel das Feld 'bewilligungErteilt'. 

In [13]:
type(data["bewilligungenErteilt"]) == list

True

In [12]:
output = {}
for k,v in data.items(): #k für Key, v = value #.items holt k und v
    if type(v) == list: #wenn bei Value eine Liste kommt, dann bearbeite das wie folgt:
        dummy_keys = []
        for i in range(0,3): #wir gehen mal davon aus, dass nicht mehr als 4 Listen enthalten sind. Muss evlt. angepasst werden
            try:
                element = v[i]
                if len(dummy_keys) > 0:
                    dummy_keys = []
                if type(element) == int:
                    #print("been here int")
                    output[k] = element
                else:
                    #print("been here")
                    for k_inside, v_inside in element.items():
                        dummy_keys.append(k_inside)
                        #print(dummy_keys)
                        if type(v_inside) == type([]):
                            for inner_element in v_inside:
                                for k_inside_inside,v_inside_inside in inner_element.items():
                                    dummy_keys.append(k_inside+"-"+k_inside_inside)
                                    #print(dummy_keys)
                                    output[k+"-"+str(i)+"-"+k_inside+"-"+k_inside_inside] = v_inside_inside
                        else:
                            output[k+"-"+str(i)+"-"+k_inside] = v_inside
            except:
                #print(dummy_keys)
                for key in dummy_keys:
                    output[k+"-"+str(i)+"-"+ key] = ''
    else:
        #print("outer")
        output[k] = v
pd.DataFrame(output,index=[0]).columns

outer
outer
outer
outer
outer
outer
been here int
[]
[]
outer
outer
outer
outer
been here
['kantonId']
['kantonId', 'praxisname']
['kantonId', 'praxisname', 'strasse']
['kantonId', 'praxisname', 'strasse', 'plz']
['kantonId', 'praxisname', 'strasse', 'plz', 'ort']
['kantonId', 'praxisname', 'strasse', 'plz', 'ort', 'telefon1']
['kantonId', 'praxisname', 'strasse', 'plz', 'ort', 'telefon1', 'emailadresse']
['kantonId', 'praxisname', 'strasse', 'plz', 'ort', 'telefon1', 'emailadresse', 'zusatz1']
['kantonId', 'praxisname', 'strasse', 'plz', 'ort', 'telefon1', 'emailadresse', 'zusatz1', 'zusatz2']
['kantonId', 'praxisname', 'strasse', 'plz', 'ort', 'telefon1', 'emailadresse', 'zusatz1', 'zusatz2']
['kantonId', 'praxisname', 'strasse', 'plz', 'ort', 'telefon1', 'emailadresse', 'zusatz1', 'zusatz2']
[]
[]
[]
been here
['kantonId']
['kantonId', 'auflageEinschraenkungen']
['kantonId', 'auflageEinschraenkungen', 'auflageEinschraenkungen-typId']
['kantonId', 'auflageEinschraenkungen', 'auflageE

Index(['idPerson', 'name', 'vorname', 'geschlechtId', 'geburtsdatum',
       'jahrgang', 'sprachkompetenzIds', 'nationalitaet1Id',
       'nationalitaet2Id', 'gln', 'uid', 'adressenBewilligungen-0-kantonId',
       'adressenBewilligungen-0-praxisname', 'adressenBewilligungen-0-strasse',
       'adressenBewilligungen-0-plz', 'adressenBewilligungen-0-ort',
       'adressenBewilligungen-0-telefon1',
       'adressenBewilligungen-0-emailadresse',
       'adressenBewilligungen-0-zusatz1', 'adressenBewilligungen-0-zusatz2',
       'adressenBewilligungen-1-kantonId',
       'adressenBewilligungen-1-praxisname', 'adressenBewilligungen-1-strasse',
       'adressenBewilligungen-1-plz', 'adressenBewilligungen-1-ort',
       'adressenBewilligungen-1-telefon1',
       'adressenBewilligungen-1-emailadresse',
       'adressenBewilligungen-1-zusatz1', 'adressenBewilligungen-1-zusatz2',
       'adressenBewilligungen-2-kantonId',
       'adressenBewilligungen-2-praxisname', 'adressenBewilligungen-2-stra

In [23]:
pd.DataFrame(output,index=[0]).T

0
idPerson                                    11036
name                            Schweizer-Vüllers
vorname                                   Andreas
geschlechtId                                12000
geburtsdatum                                 None
...                                           ...
abschluesse-2-organisationName                   
abschluesse-2-organisationPlz                    
abschluesse-2-organisationOrt                    
canDoMutationsantrag                        False
canHaveBewilligungen                         True

[99 rows x 1 columns]

Nochmals, Schritt für Schritt (Unterricht am 9.12.2019):

In [8]:
for k,v in data.items():
    print(str(k), str(v))

idPerson 1
name Birchler Pedross
vorname Angelina
geschlechtId 12001
geburtsdatum None
jahrgang 1968
sprachkompetenzIds [13000]
nationalitaet1Id 10001
nationalitaet2Id None
gln 7601003917000
uid None
adressenBewilligungen [{'kantonId': 11020, 'praxisname': '-', 'strasse': 'Zelgstr. 5', 'plz': '8003', 'ort': 'Zürich', 'telefon1': '', 'emailadresse': '', 'zusatz1': '', 'zusatz2': ''}]
adressenNeunzigTageDienstleister []
bewilligungenErteilt [{'kantonId': 11020, 'auflageEinschraenkungen': [{'typId': 8003, 'datumEntscheid': '2010-08-02T00:00:00', 'datumBefristung': '2020-08-31T00:00:00'}], 'datumEntscheid': '2010-08-02T00:00:00', 'datumErteilung': '2018-10-23T15:12:51.923', 'datumBefristung': '2018-10-30T07:07:15.607'}]
bewilligungenEntzogen []
bewilligungenVerweigert []
bewilligungenKeineBewilligung []
neunzigTageDienstleister []
abschluesse [{'berufsbezeichnungId': 54001, 'abschlussId': 55001, 'ausstellungsdatum': '2001-01-01T00:00:00', 'ausstellungslandId': 10001, 'anerkennungstypId': N

Das Problem: in adressenBewilligungen zum Beispiel haben wir eine Liste drin. 

In [19]:
liste = []
minidict = {}

for k,v in data.items():
    if type(v) == list:
        for i in range(0,3):
            try:
                print('juhu, hier ist was')
                minidict[k] = v[i]
            except: 
                print('Mist, Element gibt es nicht')
        print('Das war jetzt gerade die Extrawurst (Liste)')
    else:
        print('Alles gut')
        minidict[k] = v
liste.append(minidict)
pd.DataFrame.from_dict(liste)


Alles gut
Alles gut
Alles gut
Alles gut
Alles gut
Alles gut
juhu, hier ist was
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier ist was
Mist, Element gibt es nicht
Das war jetzt gerade die Extrawurst (Liste)
Alles gut
Alles gut
Alles gut
Alles gut
juhu, hier ist was
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier ist was
Mist, Element gibt es nicht
Das war jetzt gerade die Extrawurst (Liste)
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier ist was
Mist, Element gibt es nicht
Das war jetzt gerade die Extrawurst (Liste)
juhu, hier ist was
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier ist was
Mist, Element gibt es nicht
Das war jetzt gerade die Extrawurst (Liste)
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier ist was
Mist, Element gibt es nicht
Das war jetzt gerade die Extrawurst (Liste)
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier i

idPerson              name   vorname  geschlechtId geburtsdatum  jahrgang  \
0         1  Birchler Pedross  Angelina         12001         None      1968   

   sprachkompetenzIds  nationalitaet1Id nationalitaet2Id            gln   uid  \
0               13000             10001             None  7601003917000  None   

                               adressenBewilligungen  \
0  {'kantonId': 11020, 'praxisname': '-', 'strass...   

                                bewilligungenErteilt  \
0  {'kantonId': 11020, 'auflageEinschraenkungen':...   

                                         abschluesse  canDoMutationsantrag  \
0  {'berufsbezeichnungId': 54002, 'abschlussId': ...                 False   

   canHaveBewilligungen  
0                  True

In [20]:
pd.DataFrame.from_dict(liste).T

0
idPerson                                                               1
name                                                    Birchler Pedross
vorname                                                         Angelina
geschlechtId                                                       12001
geburtsdatum                                                        None
jahrgang                                                            1968
sprachkompetenzIds                                                 13000
nationalitaet1Id                                                   10001
nationalitaet2Id                                                    None
gln                                                        7601003917000
uid                                                                 None
adressenBewilligungen  {'kantonId': 11020, 'praxisname': '-', 'strass...
bewilligungenErteilt   {'kantonId': 11020, 'auflageEinschraenkungen':...
abschluesse            {'berufsbezeichnungId': 54002, 'abschlussId': ...
canDoMutationsantrag                                               False
canHaveBewilligungen                                                True

Bei Abschlüsse fehlt jetzt 55001. Der Grund: es wurde beim Erstellen des Minidicts überschrieben. 
Das heisst, wir müssen neue Keys generieren. Das Problem: bei gewissen Einträgen werden die Kolonnen der Tabelle nicht mehr passen. 


In [25]:
liste = []
minidict = {}

for k,v in data.items():
    if type(v) == list:
        for i in range(0,3):
            try:
                print('juhu, hier ist was')
                minidict[k+str(i)] = v[i] #hier erstellen wir neue Keys
            except: 
                print('Mist, Element gibt es nicht')
                minidict[k+str(i)] = v[i]
        print('Das war jetzt gerade die Extrawurst (Liste)')
    else:
        print('Alles gut')
        minidict[k] = v
liste.append(minidict)
pd.DataFrame.from_dict(liste)



Alles gut
Alles gut
Alles gut
Alles gut
Alles gut
Alles gut
juhu, hier ist was
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier ist was
Mist, Element gibt es nicht
Das war jetzt gerade die Extrawurst (Liste)
Alles gut
Alles gut
Alles gut
Alles gut
juhu, hier ist was
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier ist was
Mist, Element gibt es nicht
Das war jetzt gerade die Extrawurst (Liste)
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier ist was
Mist, Element gibt es nicht
Das war jetzt gerade die Extrawurst (Liste)
juhu, hier ist was
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier ist was
Mist, Element gibt es nicht
Das war jetzt gerade die Extrawurst (Liste)
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier ist was
Mist, Element gibt es nicht
Das war jetzt gerade die Extrawurst (Liste)
juhu, hier ist was
Mist, Element gibt es nicht
juhu, hier i

idPerson              name   vorname  geschlechtId geburtsdatum  jahrgang  \
0         1  Birchler Pedross  Angelina         12001         None      1968   

   sprachkompetenzIds0  nationalitaet1Id nationalitaet2Id            gln  \
0                13000             10001             None  7601003917000   

    uid                             adressenBewilligungen0  \
0  None  {'kantonId': 11020, 'praxisname': '-', 'strass...   

                               bewilligungenErteilt0  \
0  {'kantonId': 11020, 'auflageEinschraenkungen':...   

                                        abschluesse0  \
0  {'berufsbezeichnungId': 54001, 'abschlussId': ...   

                                        abschluesse1  canDoMutationsantrag  \
0  {'berufsbezeichnungId': 54002, 'abschlussId': ...                 False   

   canHaveBewilligungen  
0                  True

pd.DataFrame.from_dict(liste).T

In [23]:
pd.DataFrame.from_dict(liste).T

0
idPerson                                                                1
name                                                     Birchler Pedross
vorname                                                          Angelina
geschlechtId                                                        12001
geburtsdatum                                                         None
jahrgang                                                             1968
sprachkompetenzIds0                                                 13000
nationalitaet1Id                                                    10001
nationalitaet2Id                                                     None
gln                                                         7601003917000
uid                                                                  None
adressenBewilligungen0  {'kantonId': 11020, 'praxisname': '-', 'strass...
bewilligungenErteilt0   {'kantonId': 11020, 'auflageEinschraenkungen':...
abschluesse0            {'berufsbezeichnungId': 54001, 'abschlussId': ...
abschluesse1            {'berufsbezeichnungId': 54002, 'abschlussId': ...
canDoMutationsantrag                                                False
canHaveBewilligungen                                                 True

Fazit: immer wieder printen. 

In [26]:
liste = []
minidict = {}
for k,v in data.items():
    print("Ich bin gerade bei Key: " + str(k))
    print("Das dazugehörige element ist: " + str(v))
    if type(v) == list:
        for i in range(0,3):
            try:
                print("Juhu ich hab wenigstens was gefunden!")
                minidict[k+str(i)] = v[i]
                print("Minidict hat gerade drin:" + str(minidict))
            except:
                minidict[k+str(i)] = "wurst"
                print("mist, element gibts nicht...")
        print("extrawurst")
    else:
        print("alles gut")
        minidict[k] = v
liste.append(minidict)
pd.DataFrame.from_dict(liste).T

Ich bin gerade bei Key: idPerson
Das dazugehörige element ist: 1
alles gut
Ich bin gerade bei Key: name
Das dazugehörige element ist: Birchler Pedross
alles gut
Ich bin gerade bei Key: vorname
Das dazugehörige element ist: Angelina
alles gut
Ich bin gerade bei Key: geschlechtId
Das dazugehörige element ist: 12001
alles gut
Ich bin gerade bei Key: geburtsdatum
Das dazugehörige element ist: None
alles gut
Ich bin gerade bei Key: jahrgang
Das dazugehörige element ist: 1968
alles gut
Ich bin gerade bei Key: sprachkompetenzIds
Das dazugehörige element ist: [13000]
Juhu ich hab wenigstens was gefunden!
Minidict hat gerade drin:{'idPerson': 1, 'name': 'Birchler Pedross', 'vorname': 'Angelina', 'geschlechtId': 12001, 'geburtsdatum': None, 'jahrgang': 1968, 'sprachkompetenzIds0': 13000}
Juhu ich hab wenigstens was gefunden!
mist, element gibts nicht...
Juhu ich hab wenigstens was gefunden!
mist, element gibts nicht...
extrawurst
Ich bin gerade bei Key: nationalitaet1Id
Das dazugehörige element 

0
idPerson                                                                           1
name                                                                Birchler Pedross
vorname                                                                     Angelina
geschlechtId                                                                   12001
geburtsdatum                                                                    None
jahrgang                                                                        1968
sprachkompetenzIds0                                                            13000
sprachkompetenzIds1                                                            wurst
sprachkompetenzIds2                                                            wurst
nationalitaet1Id                                                               10001
nationalitaet2Id                                                                None
gln                                                                    7601003917000
uid                                                                             None
adressenBewilligungen0             {'kantonId': 11020, 'praxisname': '-', 'strass...
adressenBewilligungen1                                                         wurst
adressenBewilligungen2                                                         wurst
adressenNeunzigTageDienstleister0                                              wurst
adressenNeunzigTageDienstleister1                                              wurst
adressenNeunzigTageDienstleister2                                              wurst
bewilligungenErteilt0              {'kantonId': 11020, 'auflageEinschraenkungen':...
bewilligungenErteilt1                                                          wurst
bewilligungenErteilt2                                                          wurst
bewilligungenEntzogen0                                                         wurst
bewilligungenEntzogen1                                                         wurst
bewilligungenEntzogen2                                                         wurst
bewilligungenVerweigert0                                                       wurst
bewilligungenVerweigert1                                                       wurst
bewilligungenVerweigert2                                                       wurst
bewilligungenKeineBewilligung0                                                 wurst
bewilligungenKeineBewilligung1                                                 wurst
bewilligungenKeineBewilligung2                                                 wurst
neunzigTageDienstleister0                                                      wurst
neunzigTageDienstleister1                                                      wurst
neunzigTageDienstleister2                                                      wurst
abschluesse0                       {'berufsbezeichnungId': 54001, 'abschlussId': ...
abschluesse1                       {'berufsbezeichnungId': 54002, 'abschlussId': ...
abschluesse2                                                                   wurst
canDoMutationsantrag                                                           False
canHaveBewilligungen                                                            True